### PR TITLE
add a first iteration to get syntax in FHE

### PR DESCRIPTION
**Pull request by goens**

### CODE REVIEW

**Comment by bollu** [comment]
@goens we should make CI green on this one.

**Comment by goens** [comment]
@bollu yeah, we should! I am on holiday until April 15 though, so I won't be able to get to this until then.

**Comment by goens** [comment]
@bollu the FHE syntax part works now, there's random errors that I don't know how they got here (CI in main is also not green)

File: SSA/Projects/FullyHomomorphicEncryption/Syntax.lean [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+                            (@SubNegZeroMonoid.toNegZeroClass.{0} (ZMod q)
+                              (@SubtractionMonoid.toSubNegZeroMonoid.{0} (ZMod q)
+                                (@AddGroup.toSubtractionMonoid.{0} (ZMod q)
+                                  (@AddGroupWithOne.toAddGroup.{0} (ZMod q)
+                                    (@Ring.toAddGroupWithOne.{0} (ZMod q)
+                                      (@CommRing.toRing.{0} (ZMod q) (ZMod.commRing q))))))))))
+                    a)  = (0 : ZMod q) := by simp
+    set_option pp.all true in
+    try rw [heta] -- literally copy-pasted!
+    sorry
```
Comment (bollu) [inline SSA/Projects/FullyHomomorphicEncryption/Syntax.lean:56]: I am a tad worry about the `sorry`.

**Review by bollu — COMMENTED** [review]
LGTM, in that I'd suggest we merge this in and iterate in-tree.

@goens , could you help me recollect what the `sorry`s are for? I feel it's a little high priority to triage these.

**Comment by bollu** [comment]
@alexkeizer  / @tobiasgrosser , do I have the go-ahead to merge?

File: SSA/Projects/FullyHomomorphicEncryption/Syntax.lean (hunk 1 above)
Comment (goens) [reply SSA/Projects/FullyHomomorphicEncryption/Syntax.lean:56]: that's your hack to get around the noncomputable stuff leaking into the elaborator 😅

File: SSA/Projects/FullyHomomorphicEncryption/Syntax.lean [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+  case neg h =>
+    sorry
```
[the remark this answers is not in the document]
Comment (goens) [reply SSA/Projects/FullyHomomorphicEncryption/Syntax.lean:58]: that's your hack to get around the noncomputable stuff leaking into the elaborator 😅

[the remark this answers is not in the document]
Comment (bollu) [reply SSA/Projects/FullyHomomorphicEncryption/Syntax.lean:58]: *sigh* and the chickens come home to roost...

**Comment by alexkeizer** [comment]
Feel free to merge, but please then also merge into the side-effects branch, and make sure that still builds, too

**Comment by bollu** [comment]
Had a quick review on a remote call with @tobiasgrosser . The feedback was to extract out the obvious stuff from the file, and leave the hard part (R.ofZ) as a separate PR. I see the following smaller PRs:


1. Changing the signature of `R.monomial` to make `q` explicit in monomial
2. Changing `OpDenote` into an `instance FHEOpDenote `
3. Adding `fhe_com` and its associated meta code.
4. Adding `R.ofZ` and the surrounding hacks / ceremony.

**Comment by bollu** [comment]
@goens let's colab on splitting the PR? I'll do the first one one.

**Comment by bollu** [comment]
This PR is currently being broken up into pieces by Sid, to be upstreamed piecemeal.

File: SSA/Experimental/Bugred/irreducible_def_1.lean [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import Mathlib.Tactic.IrreducibleDef
+
+namespace Irred
+irreducible_def foo (x : Nat) : Int := x
+
+def bar (x : Nat) : Int := foo x
+end Irred
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Experimental/Bugred/irreducible_def_1.lean]: I dropped these following a conversation with @bollu. These are just experiments he did for himself and do not need to be in the codebase.

File: SSA/Experimental/Bugred/irreducible_def_2.lean [hunk 4]
```diff
@@ -0,0 +1 @@
+import IrreducibleDef1
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Experimental/Bugred/irreducible_def_2.lean]: Dropped.

File: SSA/Projects/MLIRSyntax/AST.lean [hunk 5]
```diff
@@ -181,6 +181,9 @@ instance : Coe String (AttrValue φ) where
 instance : Coe Int (AttrValue φ) where
   coe (i: Int) := AttrValue.int i (MLIRType.int .Signless 64)
 
+instance : Coe Nat (AttrValue φ) where
+  coe (n: Nat) := AttrValue.nat n
+
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Projects/MLIRSyntax/AST.lean]: These are not needed. I dropped them to keep this PR focused.

**Pull request closed without merging**